### ISSUE TITLE
DataGrid big table

### ISSUE THREAD

**Issue by mastras23** [dbgate/dbgate#962]
I have large partitioned tables, they cannot be opened in DataGrid despite the fact that I have a small rows limit set in the settings.
Is there a way to append a clause to individual tables that DataGrid will attach to each query?
For example, in this case, such a query works quickly because it limits the partitions:

```
select top 1000 *
from TrafficVolume
where time> DATEADD(day, -1, sysutcdatetime())
```

Table:
```
TABLE [dbo].[TrafficVolume](
	[TreeId] [uniqueidentifier] NOT NULL,
	[Time] [datetime2](7) NOT NULL,
	[Minute] [int] NOT NULL,
	[Value] [int] NOT NULL,
 CONSTRAINT [PK_TrafficVolume] PRIMARY KEY CLUSTERED 
(
	[TreeId] ASC,
	[Time] ASC,
) ON [PS_TrafficVolume]([Time])
) ON [PS_TrafficVolume]([Time])
```

**Comment by janproch** [comment]
Could you please test how quick/slow are flowwing queries?

SELECT COUNT(*) FROM [dbo].[TrafficVolume]

SELECT COUNT(1) FROM [dbo].[TrafficVolume]

select * from TrafficVolume order by TreeId offset 10 rows fetch next 10 rows only

DataGrid uses OFFSET - FETCH NEXT syntax for quering range of data and COUNT(*) query for obtaining row count

We could add some settings for big tables, to minimize queries on them

Thanks

**Comment by mastras23** [comment]
Please note that PK contains two columns {id,time}, but partitioning is by {time}, so the **query select * from TrafficVolume order by TreeId** ... will practically not be executed.
For example, I have 60 partitions of 1 month, each partition contains 20 million rows, so without narrowing it down to one partition it is impossible to execute the query in a reasonable time.
It seems to me that such functionality cannot be implemented automatically based on the description of tables from the master database.
However, if for each table the application could remember some condition defined by the user that would be added to existing queries, it would be very useful

**Comment by janproch** [comment]
OK, this is good point

These queries will be OK?
-- select TOP without condition on partitied column
select top 1000 * from TrafficVolume

-- number of records
SELECT COUNT(*) FROM [dbo].[TrafficVolume]

We would like handle this somehow (probably something like Open raw data ctx. menu), which would open table in minimal version. Or maybe per-table quick configuration, what behaviour skip in this table (eg. skip loading row count)

But I need the minimal query to obtain data, which could be deduced automatically from DB structure

Thanks

**Comment by mastras23** [comment]
are you considering an ms-sql specific or DataGrid generic solution?R.
-------- Original message --------From: Jan Prochazka ***@***.***> Date: 16/12/2024  11:42  (GMT+01:00) To: dbgate/dbgate ***@***.***> Cc: mastras23 ***@***.***>, Author ***@***.***> Subject: Re: [dbgate/dbgate] FEAT: DataGrid big table (Issue #962) 
OK, this is good point
These queries will be OK?
-- select TOP without condition on partitied column
select top 1000 * from TrafficVolume
-- number of records
SELECT COUNT(*) FROM [dbo].[TrafficVolume]
We would like handle this somehow (probably something like Open raw data ctx. menu), which would open table in minimal version. Or maybe per-table quick configuration, what behaviour skip in this table (eg. skip loading row count)
But I need the minimal query to obtain data, which could be deduced automatically from DB structure
Thanks

**Comment by janproch** [comment]
OK, thanks for the suggestion

Paging by partitions is interesting idea. I am not sure whether this is what the user primarily wants, but maybe it could be some advanced setting.

At first, I will add to roadmap to implement raw data view - minimalistic query, only SELECT TOP(N) * FROM XXX shown in grid with ability of filtering and sorting. Hopefully this query will finish, so then user could add additional filters (eg. on partitioned column time)

**Comment by mastras23** [comment]
My primary and minimal needs for this DataGrid is answer to question – are any data from today? Or what are the latest data in the table?

So, for me it will be as simple as adding clause :

where time> DATEADD(day, -1, sysutcdatetime())

**Comment by janproch** [comment]
Yes, I understand. You could write query in SQL editor tab. For use in datagrid, I need to know, whether the following query finish on partitioned table?

SELECT TOP(100) FROM [dbo].[TrafficVolume]

- only TOP clause, no ordering and no filtering

Thanks

**Comment by mastras23** [comment]
Yes, it runs quick without any sort/filter

**Comment by mastras23** [comment]
hi, 
if I open the table in raw mode, then in the datetime column filter I type the date in the form 2025-01-01 then the SQL will get =DATE() or LIKE('2024-01-01%') ?

**Comment by janproch** [comment]
Hi, 

which DB are you using?

In MS SQL, you should get something like (filtering 1965-03-03 on datetime column):

SELECT *
  FROM [dbo].[Employee] [basetbl] 
WHERE ((DATEPART(YEAR, [basetbl].[BirthDate]) = '1965') AND (DATEPART(MONTH, [basetbl].[BirthDate]) = '03') AND (DATEPART(DAY, [basetbl].[BirthDate]) = '03'))
ORDER BY [EmployeeId] ASC

**Comment by mastras23** [comment]
Hi, 

actually i had this idea in mind: 
when I type in the filter '2020-02-02' I mean to see the range
WHERE t >= 2020-02-02 00:00:00 AND t<2020-02-03 00:00:00, 

when I type in the filter '2020-02-02 10' I mean to add the clause 
WHERE t >= 2020-02-02 10:00:00 AND t<2020-02-03 11:00:00,

when I type '2020-02' in the filter I mean to add the clause 
WHERE t >= 2020-02-01 00:00:00 AND t<2020-03-01 00:00:00, 
etc.

This is intuitive and at the same time effective because the index on the column is then used and narrowing by partitioning will work. 
If you build a condition with functions e.g. DATEPART() the index will not be used and partitioning will not work - the whole data range will be scanned. 

Of course, such a filter makes sense only on the column with time, and of course you can build it in the filter expression, but such a macro would incredibly facilitate the viewing of data. 

R.
